File: v1/src/main/java/com/api/v1/doctors/domain/DoctorCrudRepository.java
package com.api.v1.doctors.domain;

import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.data.mongodb.repository.Query;

import java.util.Optional;

public interface DoctorCrudRepository extends MongoRepository<Doctor, String> {

    @Query("{ 'person.sin': ?0 }")
    Optional<Doctor> findBySIN(String sin);

    @Query("{ 'person.email': ?0 }")
    Optional<Doctor> findByEmail(String email);

    @Query("{ 'licenseNumber': ?0 }")
    Optional<Doctor> findByLicenseNumber(String licenseNumber);

    @Query("{ 'licenseNumber': ?0, 'status': 'TERMINATED' }")
    Optional<Doctor> findTerminatedByLicenseNumber(String licenseNumber);

    @Query("{ 'licenseNumber': ?0, 'status': 'ACTIVE' }")
    Optional<Doctor> findActiveByLicenseNumber(String licenseNumber);
}

File: v2/src/main/kotlin/com/api/v2/doctors/domain/DoctorCrudRepository.kt
package com.api.v2.doctors.domain

import com.api.v2.doctors.Doctor
import org.springframework.data.mongodb.repository.MongoRepository
import org.springframework.data.mongodb.repository.Query

interface DoctorCrudRepository: MongoRepository<Doctor, String> {

    @Query("{ 'licenseNumber': ?0 }")
    fun findByLicenseNumber(licenseNumber: String): Doctor?

    @Query("{ 'licenseNumber': ?0, 'status': 'ACTIVE' }")
    fun findActiveByLicenseNumber(licenseNumber: String): Doctor?

    @Query("{ 'licenseNumber': ?0, 'status': 'TERMINATED' }")
    fun findTerminatedByLicenseNumber(licenseNumber: String): Doctor?

    @Query("{ 'person.sin': ?0 }")
    fun findBySIN(sin: String): Doctor?

    @Query("{ 'person.email': ?0 }")
    fun findByEmail(email: String): Doctor?
}